content/posts/integration-test-code-coverage-in-go/index.md

```diff
@@ -0,0 +1,215 @@
+	return &killServer{
+		server: http.Server{
+			Addr: addr,
+		},
+		cancel: cancel,
+	}
+}
+
+func (s *killServer) Start() {
+	s.server.Handler = s
+
+	fmt.Println("Started KillServer")
+
+	err := s.server.ListenAndServe()
+
+	if err == http.ErrServerClosed {
+		fmt.Println("KillServer Closed")
+	} else {
+		fmt.Println("KillServer Error:", err)
+	}
+}
+
+func (s *killServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	w.WriteHeader(http.StatusOK)
+
+  // cancel the context
+	s.cancel()
+}
+
+```
+
+And our TestRun uses the context to exit once the context is canceled.
+
+```go
+func TestRun(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+  
+	killServer := newKillServer(":19999", cancel)
+	go killServer.Start()  
+	go runService(ctx)
+
+	<-ctx.Done()
+
+	killServer.server.Shutdown(context.Background())
+}
+```
+
+Next, we can compile the service with: `go test -c ./ -cover -covermode=count -coverpkg=./...` to create a `.test` binary, and run it with:
+
+```bash
+./my-service.test -test.coverprofile my-service.out
+# server started...
+```
+
+Now our server is running as it normally would, but with go's test tooling built in. All of the code it runs is counted for code coverage now.
+
+In another terminal, we run some tests and then kill it with the HTTP signal.
+
+```bash
+./run-tests.sh
+
+curl localhost:19999 # kills the server in the other terminal
+```
+
+A file named `my-service.out` should be created with coverage information from our run.
+
+
+Using `go test -c` to build test binaries is a nice little trick I learned about recently. I suppose we could accomplish the same thing by just running all our servers with the `go test` command directly, instead of compiling them into binary, but then we would need go installed in our containers.
+
+## Part 2: Running Things
+
+Now that our services are built in test mode, we can add them to our Dockerfile, with a little wrapper script to restart them endleslessly.
+
+```Dockerfile
+FROM debian
+
+WORKDIR /
+# coverage test
+COPY ./server.test /server.test
+COPY ./server_with_restart.sh /server_with_restart.sh
+
+ENTRYPOINT [ "/server_with_restart.sh" ]
+```
+
+server_with_restart.sh
+
+```sh
+#!/bin/sh
+
+while true; do
+    echo "Test server started. Built with 'go test -cover -c'"
+
+    /server.test \
+        -test.coverprofile=cover.out \
+        "$@" || exit 1; # exit if process exited with error
+    
+    # Server was probably stopped by http kill signal
+    # to collect code coverage. Restart.        
+    echo "Server restarting.."
+done
+```
+
+Now start those containers, and run your tests.
+
+
+## Part 3: Collecting Coverage
+
+Now to generate, download and merge our coverage.
+
+To trigger generation, we just have to send an HTTP request to our killserver on port 19999. Not too hard, we can just install curl in our kubernetes container, and run it from the outside with exec.
+
+```bash
+kubectl exec my-service -- curl -s localhost:19999
```

Comment:
Readers are gonna be so perplexed. Like, "where the hell did this kubernetes cluster come from!" :rofl:

Reply:
I guess we do want to use kubectl here, since it'll be more beneficial to our devs who are working on it.
But for an outsider, reading this blog is going to be confusing

Reply:
Otherwise, we'd need to explain how to install minikube and stuffs, or refactor this blog to using docker/docker-compose only.

Reply:
Just a thought, so i guess don't need to change anything here.